### ISSUE TITLE
CUDA Out of Memory Error While Training Zipformer (pruned_transducer_stateless7_streaming) on Single/Multiple H200 GPU(s).

### ISSUE THREAD

**Issue by abhay-tih** [k2-fsa/icefall#1931]
I'm encountering a CUDA Out Of Memory (OOM) issue while training `zipformer` in `icefall`, using the following configuration:

- **GPUs**: 3 x NVIDIA H200 (140 GiB memory each)
- **Training script**: `pruned_transducer_stateless7_streaming/train.py`
- **CUDA_VISIBLE_DEVICES**: 5,6,7
- **Command**:
  ```bash
  CUDA_VISIBLE_DEVICES=5,6,7 python3 pruned_transducer_stateless7_streaming/train.py \
      --exp-dir pruned_transducer_stateless7_streaming/exps/zipformer-testing-3 \
      --world-size 3 \
      --max-duration 800 \
      --num-workers 16 \
      --num-epochs 80
  ```
### System Info

  - GPU: 3 x NVIDIA H200 (140 GiB each)
  - OS / Distro: Ubuntu 24.04.2 LTS
  - CUDA Version: 12.4
  - PyTorch Version: 2.6.0

### 📊 Cut Statistics

![Image](https://github.com/user-attachments/assets/15aedea7-c2ae-44c5-ab0a-c44feec9b110)

## Issue

**Despite having ample GPU memory (3 × 140 GiB per H200 GPU), the memory usage spikes rapidly and reaches the limit, eventually causing a CUDA OOM error. This issue also occurs when training on a single GPU using the same configuration. The problem persists even when lowering the --max-duration to 600 and 400. The images below show GPU utilization just before the crash in both multi-GPU and single-GPU experiments. It’s unclear why such relatively small --max-duration values are still leading to OOM, especially on some of the most capable and memory-efficient GPUs from NVIDIA.**

![Image](https://github.com/user-attachments/assets/d6685805-c7e8-47d9-ba2d-f70e591b013e)

![Image](https://github.com/user-attachments/assets/c974f85c-8a19-4ad4-ba09-dbd0f29492ea)

#### Request for Help

Please ...can anyone advise on how to resolve or debug this memory spike. @csukuangfj @danpovey @yaozengwei 

Thanks for the amazing work on icefall!

**Comment by csukuangfj** [comment]
Please analyze the.saved pt file on error and check the max duration in the pt file.

**Comment by abhay-tih** [comment]
> Please analyze the.saved pt file on error and check the max duration in the pt file.

So before crashing out it saved a batch in a .pt file. So I analysed that .pt file and found that the maximum duration of the cut is 133.2245 secs and total duration of batch is 509.23 secs which is much lower than 800 secs (max_duration)

Any comments ? @csukuangfj

**Comment by csukuangfj** [comment]
a single wav usually has less than 30 seconds.

The longest wav in.the batch has more than 100s, which causes the OOM.error.

Please refero.to our zipformer/train.py from librispeech. It has a function to remove long utterances.

**Comment by abhay-tih** [comment]
> a single wav usually has less than 30 seconds.
> 
> The longest wav in.the batch has more than 100s, which causes the OOM.error.
> 
> Please refer.to our zipformer/train.py from librispeech. It has a function to remove long utterances.

Thanks to @csukuangfj — filtering out audios longer than 30 seconds worked!  
However, I'm now facing a new issue: during training, starting from the 2nd epoch, the losses become `NaN` and `inf`.

Here's a log snippet:

```
2025-04-25 20:45:46,067 INFO [train.py:889] (4/5) Epoch 2, batch 650, 
loss[loss=inf, simple_loss=inf, pruned_loss=inf, over 48180.00 frames.], 
tot_loss[loss=0.316, simple_loss=0.3189, pruned_loss=0.1509, over 8952492.50 frames.], 
batch size: 264, lr: 4.77e-02
```

Later, the total loss turns into an empty list:

```
2025-04-25 20:55:41,103 INFO [train.py:889] (4/5) Epoch 3, batch 100, 
loss[loss=inf, simple_loss=inf, pruned_loss=inf, over 48655.00 frames.], 
tot_loss[over 3693282.91 frames.], 
batch size: 174, lr: 4.46e-02
```

Also, after the 2nd epoch, the validation loss becomes a perfect zero.

```
2025-04-25 20:53:30,142 INFO [train.py:914] (4/5) Computing validation loss
2025-04-25 20:53:35,912 INFO [zipformer.py:2441] (4/5) attn_weights_entropy = tensor([nan, nan, nan, nan, nan, nan, nan, nan], device='cuda:4'), covar=tensor([nan, nan, nan, nan, nan, nan, nan, nan], device='cuda:4'), in_proj_covar=tensor([nan, nan, nan, nan, nan, nan, nan, nan], device='cuda:4'), out_proj_covar=tensor([nan, nan, nan, nan, nan, nan, nan, nan], device='cuda:4')
2025-04-25 20:53:37,179 INFO [train.py:923] (4/5) Epoch 3, validation: loss=0, over 2501953.00 frames.
```

I'm running with the following command:

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3,4 python3 pruned_transducer_stateless7_streaming/train.py \
  --exp-dir pruned_transducer_stateless7_streaming/exps/zipformer-testing-2000 \
  --world-size 5 \
  --max-duration 2000 \
  --num-workers 16 \
  --num-epochs 80
```

**Comment by csukuangfj** [comment]
Can you tell us what you have changed?

The default train.py should have already filtered long utterances and you don't need to.add it by yourself unless you have removed it.

Now you have inf but we don't know what else you have done to the train.py

**Comment by abhay-tih** [comment]
> Can you tell us what you have changed?
> 
> The default train.py should have already filtered long utterances and you don't need to add it by yourself unless you have removed it.
> 
> Now you have inf but we don't know what else you have done to the train.py

Yeah I only used that already implemented function of removing long utterances. The only thing I changed is that, I keep utterances of less than 1 second also.
